R version 4, featureCounts running once only, and conditional differential expression processing

yuukiiwa opened the pull request:
I have changed the docker container that contains dependencies compatible with R 4.0.0. featureCounts only runs once per sample, which I added a mergeGTFs process prior to it (please change the permission for bin/mergeGTF.py to 755 to run the pipeline). This should work if all samples use the same annotation file, which may not be the case sometimes, so I will try to figure out how to work on that. DESeq2 and DEXSeq only run when there are two or more unique conditions after checking the number of unique conditions from the sample sheet.

Here are the screenshoot of the runs:
![one_condition](https://user-images.githubusercontent.com/41866052/82772625-49cf3c80-9e72-11ea-95a3-c33ad9b3c904.png)
![two_conditions](https://user-images.githubusercontent.com/41866052/82772631-4d62c380-9e72-11ea-89c3-8de6d5b75763.png)

Thanks.

yuukiiwa committed:
Update runDESeq2.R

yuukiiwa committed:
Update main.nf

yuukiiwa committed:
Update runDESeq2.R

yuukiiwa committed:
Update main.nf

yuukiiwa committed:
Update runDESeq2.R

yuukiiwa committed:
Update runDEXseq.R

yuukiiwa committed:
Update environment.yml

yuukiiwa committed:
Update nextflow.config

yuukiiwa committed:
Update main.nf

yuukiiwa committed:
Update main.nf

yuukiiwa committed:
Update main.nf

yuukiiwa committed:
Update check_samplesheet.py

yuukiiwa committed:
Update check_samplesheet.py

yuukiiwa committed:
Update main.nf

yuukiiwa committed:
Update check_samplesheet.py

yuukiiwa committed:
Update check_samplesheet.py

yuukiiwa committed:
Update main.nf

yuukiiwa committed:
Add files via upload

yuukiiwa committed:
Update main.nf

yuukiiwa committed:
Add files via upload

yuukiiwa commented:
I tried using stringtie --merge, but it outputted a merged gtf file that is around 231 Mb, which does not have as many lines as the annotation gtf file (> 1 Gb).

yuukiiwa committed:
Update main.nf

yuukiiwa committed:
Add files via upload

yuukiiwa committed:
Added in a Bambu option (one-condition works)

yuukiiwa committed:
Update nextflow.config

yuukiiwa committed:
Update main.nf

yuukiiwa committed:
Added in the option of using Bambu count files

yuukiiwa committed:
Update main.nf

yuukiiwa committed:
Update main.nf

yuukiiwa committed:
Update runDEXseq.R

yuukiiwa committed:
Add in the transcriptquant flag

yuukiiwa commented:
I added in the flag --transcriptquant (default: bambu), which allows the user to use either stringtie (Step 2a) or bambu (Step 2b) as the transcript quantification method when running the pipeline. If bambu is being run, the downstream processes DESeq2 and DEXSeq will use the counts files for genes and transcripts generated from Bambu (this relates to the pull request I submitted for writeBambuOutput).

cying111 commented on bin/runDESeq2.R:
```diff
@@ -10,31 +10,43 @@ if (!require("DESeq2")){
+
+#sampInfo <- read.csv("~/Downloads/nanorna-bam-master/two_conditions.csv",row.names = 1)
+sampInfo<-read.csv(args[3],row.names=1)
+countTab <- data.frame(read.table("~/Downloads/nanorna-bam-master/results/Bambu/counts_gene.txt"))
```
this can be changed to the directory where bambu output saved instead of hardcoded?

yuukiiwa replied:
I have deleted that line as line 39 -- countTab <- data.frame(read.table(path,sep="\t",header=TRUE)) -- should point to the path where the count_gene.txt file is.

yuukiiwa committed:
Update runDESeq2.R

yuukiiwa committed:
add in gffcompare

yuukiiwa committed:
update the new container w/ gffcompare

yuukiiwa committed:
process replaced by gffcompare

yuukiiwa committed:
merge GTF files using gffcompare

yuukiiwa committed:
install bambu if bambu is not installed

yuukiiwa committed:
Update runBambu.R

yuukiiwa committed:
add in the option of using fasta as genomeSequence

yuukiiwa committed:
add in r-devtools

yuukiiwa committed:
change docker container with devtools

I have problem running bambu using docker even with devtools specified as a dependency. The pipeline can run locally.

yuukiiwa committed:
add --genomeSequence flag for a fasta file

yuukiiwa committed:
Update nextflow.config

yuukiiwa committed:
Add files via upload

cying111 commented on main.nf:
```diff
@@ -165,60 +170,126 @@ process StringTie2 {
      file("*.version") into ch_feat_counts_version
-     val "result/featureCounts_gene" into ch_deseq_indir
-     val "result/featureCounts_transcript" into ch_dex_indir
+     val "$baseDir/results/featureCounts_transcript" into ch_deseq2_indir
```
should it $params.outdir instead? same for the line below

yuukiiwa replied:
I have gotten problems using $params.outdir instead of $baseDir/results, which $params.outdir maps to $baseDir/work/<random_string>/.

cying111 commented on main.nf:
```diff
@@ -165,60 +170,126 @@ process StringTie2 {
+  val transcriptquant from ch_transcriptquant
+
+  output:
+  val "$baseDir/results/Bambu/counts_gene.txt" into ch_deseq2_in
```
should it $params.outdir instead? same for the line below

yuukiiwa replied:
I have gotten problems using $params.outdir instead of $baseDir/results, which $params.outdir maps to $baseDir/work/<random_string>/.

yuukiiwa committed:
genome and transcriptome inputs

bambu with the fasta file argument, which should work without the error (Error in value[[3L]](cond) :  record 1 (22:10716989-10716990) failed).

yuukiiwa committed:
transcriptome column points to annotation.gtf

yuukiiwa committed:
genome (fasta) column and transcriptome (gtf) column

yuukiiwa committed:
Update usage.md

yuukiiwa committed:
igenome fasta path uses .fasta

yuukiiwa committed:
now can use igenome keys

yuukiiwa committed:
using file instead of val for bambu inputs

yuukiiwa commented:
The updates are for the genome (fasta) and transcriptome (gtf) columns. With both fasta and gtf file paths provided or with only the igenome key provided in the genome column, the run with stringtie completed without errors. Here are the completed stringtie runs:
Run with igenome key only:
samplesheet input
![Screen Shot 2020-06-16 at 9 54 37 AM](https://user-images.githubusercontent.com/41866052/84723104-69f8a400-afb7-11ea-8916-95a4939c1821.png)
run
![Screen Shot 2020-06-16 at 9 55 24 AM](https://user-images.githubusercontent.com/41866052/84723146-872d7280-afb7-11ea-8b92-b009c15ae3ee.png)
Run with both the fasta and gtf files provided:
samplesheet input
![Screen Shot 2020-06-16 at 9 56 07 AM](https://user-images.githubusercontent.com/41866052/84723193-9f04f680-afb7-11ea-95ff-ade253774b00.png)
run 
![Screen Shot 2020-06-16 at 9 56 37 AM](https://user-images.githubusercontent.com/41866052/84723223-b17f3000-afb7-11ea-885a-d88734e4cb53.png)

yuukiiwa committed:
updating bambu to use igenome gtf and fasta files.

yuukiiwa committed:
added an argument for gtf in bambu

yuukiiwa commented:
The two new commits allows runBambu.R taking in the igenome files (fasta and gtf). The run with bambu with ftp://ftp.ensembl.org/pub/release-91/gtf/homo_sapiens/Homo_sapiens.GRCh38.91.gtf.gz worked well but not with the igenome GRCh37 gtf file (error message below).
Here is the completed bambu run with both fasta and gtf file provided (ran locally):
samplesheet input
![Screen Shot 2020-06-16 at 12 58 15 PM](https://user-images.githubusercontent.com/41866052/84733991-5a3a8900-afd2-11ea-9763-27fd3959d98e.png)
run
![Screen Shot 2020-06-16 at 12 46 28 PM](https://user-images.githubusercontent.com/41866052/84734005-6292c400-afd2-11ea-9ecb-f76088f9ef0c.png)
Here is the error message from running bambu supplying only the igenome key (samplesheet.csv used was one_igenome.csv please see above -- stringtie2 igenome key only samplesheet input):
![Screen Shot 2020-06-16 at 12 59 44 PM](https://user-images.githubusercontent.com/41866052/84734017-6d4d5900-afd2-11ea-840b-202920d33a87.png)

cying111 commented:
hi, @yuukiiwa  Is there any reason why GRCh 37 igenome is used? Maybe there is an incompatiblity issue because the alignment bam file is obtained by aligning to GRCh 38. Could you see if you can work with GRCh 38 igenome gtf file?

jonathangoeke commented:
> The two new commits allows runBambu.R taking in the igenome files (fasta and gtf). The run with bambu with ftp://ftp.ensembl.org/pub/release-91/gtf/homo_sapiens/Homo_sapiens.GRCh38.91.gtf.gz worked well but not with the igenome GRCh37 gtf file (error message below).
> Here is the completed bambu run with both fasta and gtf file provided (ran locally):
> samplesheet input
> ![Screen Shot 2020-06-16 at 12 58 15 PM](https://user-images.githubusercontent.com/41866052/84733991-5a3a8900-afd2-11ea-9763-27fd3959d98e.png)
> run
> ![Screen Shot 2020-06-16 at 12 46 28 PM](https://user-images.githubusercontent.com/41866052/84734005-6292c400-afd2-11ea-9ecb-f76088f9ef0c.png)
> Here is the error message from running bambu supplying only the igenome key (samplesheet.csv used was one_igenome.csv please see above -- stringtie2 igenome key only samplesheet input):
> ![Screen Shot 2020-06-16 at 12 59 44 PM](https://user-images.githubusercontent.com/41866052/84734017-6d4d5900-afd2-11ea-840b-202920d33a87.png)

Hi @yuukiiwa do you want to look into this in the bambu package? It seems like the prepareAnnotationsFromGTF function could be modified to prevent this error?

yuukiiwa committed:
Add in GRCh38

yuukiiwa commented:
I added GRCh38 in igenomes.config.  The igenome gtf file ran through prepareAnnotationFromGTF, but I got the same error running bambu as my previous attempts not using Homo_sapiens.GRCh38.91.gtf. Here is the error message from the run:
![Screen Shot 2020-06-16 at 4 14 37 PM](https://user-images.githubusercontent.com/41866052/84749794-08532c80-afed-11ea-824f-30410aca2875.png)
Here is the previous error message on RStudio using gencode.v33.annotation.gtf:
![sessioninfo()](https://user-images.githubusercontent.com/41866052/84750278-a810ba80-afed-11ea-9f74-9547df5a08f5.png)

yuukiiwa commented:
I submitted a pull request for bambu (https://github.com/GoekeLab/bambu/pull/116), which solves the igenome usage problem. Here is the run and the samplesheet.csv (one_igenome.csv):
![Screen Shot 2020-06-16 at 7 28 46 PM](https://user-images.githubusercontent.com/41866052/84769240-16fb0d00-b008-11ea-8e80-73a4d4c2521e.png)
Update:
I don't think the pull request above is necessary for nanorna-bam running with igenome fasta and gtf files if all the packages used by bambu are updated. I have reinstalled bambu by devtools::install("GoekeLab/bambu") with the option of updating all the packages. Here is a nanorna-bam run with the current version of bambu:
![Screen Shot 2020-06-17 at 11 20 24 AM](https://user-images.githubusercontent.com/41866052/84851330-903c4380-b08c-11ea-9bb2-9cd8c5408656.png)
